social-network-api/src/main/java/com/socialnetwork/api/controller/PostController.java

```diff
@@ -1,90 +1,137 @@
+            .stream()
+            .map(this::convertToPostDto)
+            .toList();
+  }
 
   @PostMapping()
-  public ResponseEntity<?> addPost(@RequestBody PostDto postDto) throws NoUserWithSuchCredentialsException {
-    //Retrieving user id from DTO (only needed field in request), looking for it
-    // in userservice and throwing exception if user with such id wasn`t found
-    //Also checking of JWT token and comparing it with user needed to be realise in future TODO
-    Post post = convertToPost(postDto);
-    Optional<User> user = userService.findById(post.getAuthor().getId());
-    if (user.isEmpty()) {
-      throw new NoUserWithSuchCredentialsException();
-    }
-    post.setAuthor(user.get());
+  public Response addPost(@RequestBody PostDto.Request.Created postDto, HttpServletRequest request)
+          throws NoUserWithSuchCredentialsException, NoPostWithSuchIdException, AccessDeniedException {
+    User user = userService.getReferenceById(postDto.getUser().getId());
+    Integer originalPostId = postDto.getOriginalPostId();
+    Post post = convertToPost(postDto, user, originalPostId);
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
     postService.save(post);
-    return ResponseEntity.status(201).body(new Response("Post was created"));
+    return new Response("Post was created");
   }
 
-  //  @PatchMapping()
-  //  public ResponseEntity<?> editPost(@RequestBody PostDto postDto) {
-  //    Post post = convertToPost(postDto);
-  //    if (!postService.existsById(post.getPostId())) {
-  //      return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new Response(POST_NOT_FOUND));
-  //    }
-  //    postService.edit(post);
-  //    return ResponseEntity.ok(new Response("Post was edited successfully"));
-  //  }
-
-  //  @DeleteMapping()
-  //  public ResponseEntity<?> deletePost(@RequestBody PostDto postDto) {
-  //    Post post = convertToPost(postDto);
-  //    if (!postService.existsById(post.getPostId())) {
-  //      return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new Response(POST_NOT_FOUND));
-  //    }
-  //    postService.delete(post);
-  //    return ResponseEntity.ok(new Response("Post was deleted"));
-  //  }
-
-  private Post convertToPost(PostDto postDto) {
-    return modelMapper.map(postDto, Post.class);
+  @PostMapping("/bookmark")
+  public PostDto.Response.Default
+    saveBookmark(@RequestBody PostDto.Request.Action postDto, HttpServletRequest request)
+          throws NoUserWithSuchCredentialsException, NoPostWithSuchIdException, AccessDeniedException {
+    User user = userService.getReferenceById(postDto.getUser().getId());
+    Post post = postService.getReferenceById(postDto.getPost().getId());
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
+
+    if (bookmarkService.existsByUserAndPost(user, post)) {
```

Comment:
This logic should be in service. Call for method 'switch' only in the controller, for example. And reduce excessive object creation.